File: Example/QDNavigationBar/QDConfigEditor/QDColorPickerController.m
//
//  QDColorPickerController.m
//  QDCustomNavigation_Example
//
//  Created by Sinno on 2020/8/5.
//  Copyright © 2020 sinno93. All rights reserved.
//

#import "QDColorPickerController.h"
#import <Masonry/Masonry.h>
#import "QDColorSliderView.h"

@interface QDColorPickerController ()

@property (nonatomic, strong) UIView *containerView;
@property (nonatomic, strong) UIView *colorView;
@property (nonatomic, strong) UIStackView *stackView;
@property (nonatomic, strong) QDColorSliderView *redSliderView;
@property (nonatomic, strong) QDColorSliderView *greenSliderView;
@property (nonatomic, strong) QDColorSliderView *blueSliderView;
@property (nonatomic, strong) QDColorSliderView *alphaSliderView;

@property (nonatomic, strong) UIButton *closeButton;
@property (nonatomic, strong) UILabel *titleLabel;

@end

@implementation QDColorPickerController



#pragma mark - Initialization
- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        self.modalPresentationStyle = UIModalPresentationOverFullScreen;
    }
    return self;
}
#pragma mark - Life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    [self configSubviews];
    [self setData];
    self.containerView.hidden = YES;
    // Do any additional setup after loading the view.
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self show];
}

#pragma mark - Actions
- (void)onColorChanged {
    CGFloat red = self.redSliderView.value;
    CGFloat green = self.greenSliderView.value;
    CGFloat blue = self.blueSliderView.value;
    CGFloat alpha = self.alphaSliderView.value;
    
    UIColor *color = [UIColor colorWithRed:red green:green blue:blue alpha:alpha];
    _color = color;
    [self setData];
    if (self.colorChanged) {
        self.colorChanged(color);
    }
}

- (void)closeButtonClick:(UIButton *)button {
    [self dismiss];
}

- (void)emptyTap:(UITapGestureRecognizer *)tap {
    CGPoint point = [tap locationInView:self.containerView];
    
    if (!CGRectContainsPoint(self.containerView.bounds, point)) {
        [self dismiss];
    }
    
}

#pragma mark - Delegates
#pragma mark - Public methods
- (void)show {
    self.view.backgroundColor = [UIColor.blackColor colorWithAlphaComponent:0];
    [self.view layoutIfNeeded];
    self.containerView.hidden = NO;
    self.containerView.transform = CGAffineTransformMakeTranslation(0, self.containerView.frame.size.height);
    [UIView animateWithDuration:0.35 animations:^{
        self.containerView.transform = CGAffineTransformIdentity;
        self.view.backgroundColor = [UIColor.blackColor colorWithAlphaComponent:0.4];
    }];
}

- (void)dismiss {
    [UIView animateWithDuration:0.35 animations:^{
        self.containerView.transform = CGAffineTransformMakeTranslation(0, self.containerView.frame.size.height);
        self.view.backgroundColor = [UIColor.blackColor colorWithAlphaComponent:0];
    } completion:^(BOOL finished) {
        [self dismissViewControllerAnimated:NO completion:nil];
    }];
}

- (void)setColor:(UIColor *)color {
    _color = color;
    [self setData];
}
- (void)setData {
    UIColor *color = self.color;
    if (!color) {
        color = UIColor.whiteColor;
    }
    CGFloat red = 0;
    CGFloat green = 0;
    CGFloat blue = 0;
    CGFloat alpha = 0;
    [color getRed:&red green:&green blue:&blue alpha:&alpha];
    self.redSliderView.value = red;
    self.greenSliderView.value = green;
    self.blueSliderView.value = blue;
    self.alphaSliderView.value = alpha;
    
//    self.redSliderView.colors =
    self.redSliderView.colors = @[colorWithRGB(0, green, blue), colorWithRGB(1, green, blue)];
    self.greenSliderView.colors = @[colorWithRGB(red, 0, blue), colorWithRGB(red, 1, blue)];
    self.blueSliderView.colors = @[colorWithRGB(red, green, 0), colorWithRGB(red, green, 1)];
    self.alphaSliderView.colors = @[colorWithRGBA(0, 0, 0, 0),colorWithRGBA(0, 0, 0, 1)];
    self.colorView.backgroundColor = color;
    
    // https://blog.walterlv.com/post/get-gray-reversed-color.html
    // 计算反差色
    UIColor *contrastColor = nil;
    CGFloat grayLevel = (0.299 * red + 0.587 * green + 0.114 * blue);
    if (grayLevel > 0.5 || alpha<0.5) {
        contrastColor = UIColor.blackColor;
    } else {
        contrastColor = UIColor.whiteColor;
    }
    self.titleLabel.textColor = contrastColor;
    [self.closeButton setTitleColor:contrastColor forState:UIControlStateNormal];
    NSString *str = [QDColorPickerController hexStringFromColor:color];
    self.titleLabel.text = str;
}
UIColor *colorWithRGB(CGFloat r, CGFloat g, CGFloat b) {
    return colorWithRGBA(r, g, b, 1);
}
UIColor *colorWithRGBA(CGFloat r, CGFloat g, CGFloat b, CGFloat a) {
    return [UIColor colorWithRed:r green:g blue:b alpha:a];
}
- (UIColor *)colorWithRGBA:(NSArray *)array {
    CGFloat red = [array[0] doubleValue];
    CGFloat green = [array[1] doubleValue];;
    CGFloat blue = [array[2] doubleValue];;
    CGFloat alpha = [array[3] doubleValue];;
    return [UIColor colorWithRed:red green:green blue:blue alpha:alpha];
}
#pragma mark - Private methods
- (void)configSubviews {
    self.view.backgroundColor = UIColor.clearColor;
    [self.view addSubview:self.containerView];
    
    [self.containerView mas_makeConstraints:^(MASConstraintMaker *make) {
        if (@available(iOS 11.0, *)) {
            make.left.equalTo(self.view.mas_safeAreaLayoutGuideLeft);
            make.right.equalTo(self.view.mas_safeAreaLayoutGuideRight);
            make.bottom.equalTo(self.view);
        } else {
            make.leading.trailing.equalTo(self.view);
            make.bottom.equalTo(self.view);
        }
        
    }];
    UIView *bgView = [UIView new];
    bgView.backgroundColor = [UIColor whiteColor];
    [self.containerView addSubview:bgView];
    [bgView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.leading.trailing.equalTo(self.view);
        make.top.bottom.equalTo(self.containerView);
    }];
    [self.containerView addSubview:self.colorView];
    [self.containerView addSubview:self.closeButton];
    [self.containerView addSubview:self.stackView];
    [self.colorView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.containerView);
        make.leading.trailing.equalTo(self.view);
        make.height.equalTo(@44);
    }];
    [self.closeButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self.containerView.mas_right).offset(-15);
        
        make.centerY.equalTo(self.colorView);
    }];
    [self.stackView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.leading.trailing.equalTo(self.containerView);
        make.top.equalTo(self.colorView.mas_bottom).offset(15);
        if (@available(iOS 11.0, *)) {
            make.bottom.equalTo(self.containerView.mas_safeAreaLayoutGuideBottom).offset(-15);
        } else {
            make.bottom.equalTo(self.containerView).offset(-15);
        }
        
    }];
    
    NSArray *array = @[self.redSliderView, self.greenSliderView, self.blueSliderView, self.alphaSliderView];
    for (QDColorSliderView *view in array) {
        [self.stackView addArrangedSubview:view];
        __weak __typeof(self)weakSelf = self;
        view.valueChanged = ^(CGFloat value) {
            [weakSelf onColorChanged];
        };
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.height.equalTo(@40);
        }];
    }
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(emptyTap:)];
    [self.view addGestureRecognizer:tap];
}
#pragma mark - Setter && Getter

- (UIView *)containerView {
    if (!_containerView) {
        UIView *view = [[UIView alloc] init];
        view.backgroundColor = UIColor.whiteColor;
        _containerView = view;
    }
    return _containerView;
}

- (UIStackView *)stackView {
    if (!_stackView) {
        UIStackView *view = [[UIStackView alloc] init];
        view.axis = UILayoutConstraintAxisVertical;
        view.alignment = UIStackViewAlignmentFill;
        view.spacing = 1.0f;
        _stackView = view;
    }
    return _stackView;
}

- (UIView *)colorView {
    if (!_colorView) {
        UIView *view = [[UIView alloc] init];
        view.layer.borderColor = UIColor.grayColor.CGColor;
        view.layer.borderWidth = 1.0/UIScreen.mainScreen.scale;
        [view addSubview:self.titleLabel];
        [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.center.equalTo(view);
        }];
        _colorView = view;
    }
    return _colorView;
}

- (QDColorSliderView *)redSliderView {
    if (!_redSliderView) {
        QDColorSliderView *sliderView = [QDColorSliderView new];
        sliderView.title = @"Red";
        _redSliderView = sliderView;
    }
    return _redSliderView;
}

- (QDColorSliderView *)greenSliderView {
    if (!_greenSliderView) {
        QDColorSliderView *sliderView = [QDColorSliderView new];
        sliderView.title = @"Green";
        _greenSliderView = sliderView;
    }
    return _greenSliderView;
}

- (QDColorSliderView *)blueSliderView {
    if (!_blueSliderView) {
        QDColorSliderView *sliderView = [QDColorSliderView new];
        sliderView.title = @"Blue";
        _blueSliderView = sliderView;
    }
    return _blueSliderView;
}

- (QDColorSliderView *)alphaSliderView {
    if (!_alphaSliderView) {
        QDColorSliderView *sliderView = [QDColorSliderView new];
        sliderView.title = @"Alpha";
        _alphaSliderView = sliderView;
    }
    return _alphaSliderView;
}


- (UIButton *)closeButton {
    if (!_closeButton) {
        UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
        [button setTitle:@"确定" forState:UIControlStateNormal];
        [button setTitleColor:UIColor.systemBlueColor forState:UIControlStateNormal];
        [button addTarget:self action:@selector(closeButtonClick:) forControlEvents:UIControlEventTouchUpInside];
        _closeButton = button;
    }
    return _closeButton;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        UILabel *label = [[UILabel alloc] init];
        label.text = @"颜色选择器";
        _titleLabel = label;
    }
    return _titleLabel;
}

+ (NSString *)hexStringFromColor:(UIColor *)color {
    CGColorSpaceModel colorSpace = CGColorSpaceGetModel(CGColorGetColorSpace(color.CGColor));
    const CGFloat *components = CGColorGetComponents(color.CGColor);

    CGFloat r, g, b, a;

    if (colorSpace == kCGColorSpaceModelMonochrome) {
        r = components[0];
        g = components[0];
        b = components[0];
        a = components[1];
    } else if (colorSpace == kCGColorSpaceModelRGB) {
        r = components[0];
        g = components[1];
        b = components[2];
        a = components[3];
    }

    return [NSString stringWithFormat:@"#%02lX%02lX%02lX%02lX",
            lroundf(r * 255),
            lroundf(g * 255),
            lroundf(b * 255),
            lroundf(a * 255)];
}
@end

File: Example/QDNavigationBar/QDConfigEditor/QDColorPickerController.swift
//
//  QDColorPickerController.swift
//  QDNavigationBar_Example
//
//  Created by zqd on 2024/6/5.
//  Copyright © 2024 CocoaPods. All rights reserved.
//

import UIKit
import SnapKit

class QDColorPickerController: UIViewController {

    private var containerView: UIView = {
        let view = UIView()
        view.backgroundColor = .white
        return view
    }()
    
    private lazy var colorView: UIView = {
        let view = UIView()
        view.layer.borderColor = UIColor.gray.cgColor
        view.layer.borderWidth = 1.0 / UIScreen.main.scale
        view.addSubview(self.titleLabel)
        self.titleLabel.snp.makeConstraints { make in
            make.center.equalTo(view)
        }
        
        return view
    }()
    
    private var stackView: UIStackView = {
        let stackView = UIStackView()
        stackView.axis = .vertical
        stackView.alignment = .fill
        stackView.spacing = 1.0
        return stackView
    }()
    
    private lazy var redSliderView = {
        let view = QDColorSliderView()
        view.title = "Red"
        return view
    }()
    
    private var greenSliderView = {
        let view = QDColorSliderView()
        view.title = "Green"
        return view
    }()
    
    private var blueSliderView = {
        let view = QDColorSliderView()
        view.title = "Blue"
        return view
    }()
    
    private var alphaSliderView = {
        let view = QDColorSliderView()
        view.title = "Alpha"
        return view
    }()

    private lazy var closeButton: UIButton = {
        let button = UIButton(type: .system)
        button.setTitle("确定", for: .normal)
        button.setTitleColor(.systemBlue, for: .normal)
        button.addTarget(self, action: #selector(closeButtonClick(_:)), for: .primaryActionTriggered)
        return button
    }()
    
    private var titleLabel: UILabel = {
        let label = UILabel()
        label.text = "颜色选择器"
        return label
    }()
    
    var colorChanged: ((UIColor) -> Void)?
    var color: UIColor? {
        didSet {
            setData()
        }
    }

    // MARK: - Initialization

    init() {
        super.init(nibName: nil, bundle: nil)
        self.modalPresentationStyle = .overFullScreen
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        configSubviews()
        setData()
        containerView.isHidden = true
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        show()
    }

    // MARK: - Actions

    @objc private func onColorChanged() {
        let red = redSliderView.value
        let green = greenSliderView.value
        let blue = blueSliderView.value
        let alpha = alphaSliderView.value
        
        let color = UIColor(red: red, green: green, blue: blue, alpha: alpha)
        self.color = color
        setData()
        colorChanged?(color)
    }

    @objc private func closeButtonClick(_ button: UIButton) {
        dismiss()
    }

    @objc private func emptyTap(_ tap: UITapGestureRecognizer) {
        let point = tap.location(in: containerView)
        if !containerView.bounds.contains(point) {
            dismiss()
        }
    }

    // MARK: - Public Methods

    func show() {
        view.backgroundColor = UIColor.black.withAlphaComponent(0)
        view.layoutIfNeeded()
        containerView.isHidden = false
        containerView.transform = CGAffineTransform(translationX: 0, y: containerView.frame.size.height)
        UIView.animate(withDuration: 0.35) {
            self.containerView.transform = .identity
            self.view.backgroundColor = UIColor.black.withAlphaComponent(0.4)
        }
    }

    func dismiss() {
        UIView.animate(withDuration: 0.35, animations: {
            self.containerView.transform = CGAffineTransform(translationX: 0, y: self.containerView.frame.size.height)
            self.view.backgroundColor = UIColor.black.withAlphaComponent(0)
        }) { finished in
            self.dismiss(animated: false, completion: nil)
        }
    }

    private func setData() {
        guard let color = self.color else {
            return
        }
        
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        redSliderView.value = red
        greenSliderView.value = green
        blueSliderView.value = blue
        alphaSliderView.value = alpha
        
        redSliderView.colors = [colorWithRGB(0, green, blue), colorWithRGB(1, green, blue)]
        greenSliderView.colors = [colorWithRGB(red, 0, blue), colorWithRGB(red, 1, blue)]
        blueSliderView.colors = [colorWithRGB(red, green, 0), colorWithRGB(red, green, 1)]
        alphaSliderView.colors = [colorWithRGBA(0, 0, 0, 0), colorWithRGBA(0, 0, 0, 1)]
        colorView.backgroundColor = color
        
        // 计算反差色
        let contrastColor: UIColor
        let grayLevel = 0.299 * red + 0.587 * green + 0.114 * blue
        if grayLevel > 0.5 || alpha < 0.5 {
            contrastColor = .black
        } else {
            contrastColor = .white
        }
        titleLabel.textColor = contrastColor
        closeButton.setTitleColor(contrastColor, for: .normal)
        titleLabel.text = QDColorPickerController.hexString(from: color)
    }

    // MARK: - Private Methods

    private func configSubviews() {
        view.backgroundColor = .clear
        view.addSubview(containerView)
        
        containerView.snp.makeConstraints { make in
            if #available(iOS 11.0, *) {
                make.left.equalTo(view.safeAreaLayoutGuide.snp.left)
                make.right.equalTo(view.safeAreaLayoutGuide.snp.right)
                make.bottom.equalTo(view)
            } else {
                make.leading.trailing.equalTo(view)
                make.bottom.equalTo(view)
            }
        }
        
        let bgView = UIView()
        bgView.backgroundColor = .white
        containerView.addSubview(bgView)
        bgView.snp.makeConstraints { make in
            make.leading.trailing.equalTo(view)
            make.top.bottom.equalTo(containerView)
        }
        
        containerView.addSubview(colorView)
        containerView.addSubview(closeButton)
        containerView.addSubview(stackView)
        
        colorView.snp.makeConstraints { make in
            make.top.equalTo(containerView)
            make.leading.trailing.equalTo(view)
            make.height.equalTo(44)
        }
        
        closeButton.snp.makeConstraints { make in
            make.right.equalTo(containerView.snp.right).offset(-15)
            make.centerY.equalTo(colorView)
        }
        
        stackView.snp.makeConstraints { make in
            make.leading.trailing.equalTo(containerView)
            make.top.equalTo(colorView.snp.bottom).offset(15)
            if #available(iOS 11.0, *) {
                make.bottom.equalTo(containerView.safeAreaLayoutGuide.snp.bottom).offset(-15)
            } else {
                make.bottom.equalTo(containerView).offset(-15)
            }
        }
        
        let sliders = [redSliderView, greenSliderView, blueSliderView, alphaSliderView]
        sliders.forEach { slider in
            stackView.addArrangedSubview(slider)
            slider.snp.makeConstraints { make in
                make.height.equalTo(40)
            }
            slider.valueChanged = { [weak self] value in
                self?.onColorChanged()
            }
        }
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(emptyTap(_:)))
        view.addGestureRecognizer(tap)
    }

    private func colorWithRGB(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat) -> UIColor {
        return colorWithRGBA(r, g, b, 1)
    }

    private func colorWithRGBA(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat, _ a: CGFloat) -> UIColor {
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }

    // MARK: - Utility Methods

    private static func hexString(from color: UIColor) -> String {
        let colorSpace = color.cgColor.colorSpace!.model
        let components = color.cgColor.components!

        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0

        if colorSpace == .monochrome {
            r = components[0]
            g = components[0]
            b = components[0]
            a = components[1]
        } else if colorSpace == .rgb {
            r = components[0]
            g = components[1]
            b = components[2]
            a = components[3]
        }

        return String(format: "#%02lX%02lX%02lX%02lX",
                      lroundf(Float(r * 255)),
                      lroundf(Float(g * 255)),
                      lroundf(Float(b * 255)),
                      lroundf(Float(a * 255)))
    }
}
